### PR TITLE
[DataSource] Re-implement with flows and coroutines

### PR DESCRIPTION
**Pull request by npresseault**
## Description
Re-implement DataSource using coroutines, without needing to bridge between the 2 worlds.

## Motivation and Context
The original datasource implementation used Publishers. This one uses flows and coroutines and makes sure it behaves properly in a multithreaded usage.

## How Has This Been Tested?
The unit test from the previous implementation were migrated to the new one (using flows and coroutine). That way we can attest that both implementations have the same behaviour.

Additionally, this was tested in another real project, which used the previous datasource with publishers and validated that no behaviour changed.

## Migration from `trikot.datasources` to `trikot.datasources.streams`
`com.mirego.trikot:datasources:$trikot_version -> com.mirego.trikot:datasources-streams:$trikot_version`

## Migration from `trikot.datasources `to `trikot.datasources.flow`
Not yet documented

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by gbourassa — APPROVED** [review]

File: trikot-datasources/datasources-streams/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+```
+val myDataSource = MyDataSource<String>(MemoryCacheDataSource())
+```
+
+##### 4 - Send a request to the datasource
+
+In this example we will use memory cache
+
+```kotlin
+val request = myDataSource.myMyDataSource.read(MyDataSourceRequest(query, "queryCacheId"))
```
Comment (sdetilly) [inline trikot-datasources/datasources-streams/README.md:45]: `myDataSource.myMyDataSource` 🤔

File: trikot-datasources/datasources-streams/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    print(it)
+}
+```
+
+# Invalidation
+
+To invalidate data, simply send a get request using REFRESH_CACHE. No subscription is needed for the invalidation to happens
+
+```kotlin
+myDataSource.myMyDataSource.read(MyDataSourceRequest(query, "queryCacheId", DataSourceRequest.Type.REFRESH_CACHE))
```
Comment (sdetilly) [inline trikot-datasources/datasources-streams/README.md:56]: Same

**Review by sdetilly — APPROVED** [review]
Good job! 👍

File: trikot-datasources/datasources-streams/README.md (hunk 1 above)
Comment (remi) [reply trikot-datasources/datasources-streams/README.md]: https://open.spotify.com/track/653rxW1E7V52QWh6a7oIdS?si=b30bafce741240b3

Comment (sdetilly) [reply trikot-datasources/datasources-streams/README.md]: 🤣

**Pull request merged**